### ISSUE TITLE
Fine-Tuning: Different result between single GPU vs multi-GPU (DDP)

### ISSUE THREAD

**Issue by IamExperimenting** [huggingface/accelerate#2651]
### System Info

```Shell
accelerate verion: 0.29.1
OS: 
python version: 3.10
numpy version: 1.24.4
torch version: 2.1.0
transformer verion: 4.39.3
Accelerate configuration:
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- debug: false
	- distributed_type: MULTI_GPU
	- downcast_bf16: 'no'
	- gpu_ids: all
	- machine_rank: 0
	- main_training_function: main
	- mixed_precision: bf16
	- num_machines: 1
	- num_processes: 4
	- rdzv_backend: static
	- same_network: true
	- tpu_env: []
	- tpu_use_cluster: false
	- tpu_use_sudo: false
	- use_cpu: false

Sagemaker instance: ml.g5.12xlarge for multi-GPU , ml.g5.4xlarge for single GPU
```

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

below is the `train_ddp.py` script

```
import transformers
from trl import SFTTrainer
import torch, os, pandas as pd
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, AutoConfig
from peft import LoraConfig, AutoPeftModel, AutoPeftModelCausalLM
from datasets import Dataset, load_dataset
from accelerate import Accelerator

model_id = "google/gemma-2b"
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

tokenizer = AutoTokenizer.from_pretrained(model_id, token=os.environ['HF_TOKEN'])
model = AutoModelForCausalLM.from_pretrained(model_id, quantization_config=bnb_config, device_map={"":0}, token=os.environ['HF_TOKEN'])

data = load_dataset("Abirate/english_quotes")
data = data.map(lambda samples: tokenizer(samples["quote"]), batched=True)

lora_config = LoraConfig(
    r=8,
    target_modules=["q_proj", "o_proj", "k_proj", "v_proj", "gate_proj", "up_proj", "down_proj"],
    task_type="CAUSAL_LM",
)

def formatting_func(example):
    output_texts = []
    for i in range(len(example)):
        text = f"Quote: {example['quote'][i]}\nAuthor: {example['author'][i]}"
        output_texts.append(text)
    return output_texts

trainer = SFTTrainer(
    model=model,
    train_dataset=data["train"],
    args=transformers.TrainingArguments(
        per_device_train_batch_size=1,
        gradient_accumulation_steps=4,
        warmup_steps=2,
        max_steps=10,
        learning_rate=2e-4,
        fp16=True,
        logging_steps=1,
        output_dir="outputs",
        optim="paged_adamw_8bit"
    ),
    peft_config=lora_config,
    formatting_func=formatting_func,
)
trainer.train()

trainer.save()
```

I have provided the correct options in config file using

`accelerate config`

after setting the config file I will execute the below line to start the training.

`accelerate launch train_ddp.py`

I'm not sure where I'm making mistake, but I can see train loss is decreasing and it is almost same as the single GPU training loss.

@muellerzr @SunMarc 

Also, I cloned `https://github.com/huggingface/peft` and tried the example `https://github.com/huggingface/peft/blob/main/examples/sft/run_peft_multigpu.sh` this one is surprising showing trainable parameter as 9B, and throwing OOM error.

Additionally, I checked accelerate example, I can see only the training portion not the model saving section in the script
`https://github.com/huggingface/accelerate/blob/main/examples/nlp_example.py`

I feel I'm making some mistake while saving the model. Could you please help me to fix this?

### Expected behavior

Expecting the model to produce the same result after fine-tuning in single GPU and multi-GPU.

**Comment by muellerzr** [comment]
Per the troubleshooting guide, please check your batch size to allocate per-GPU vs actual https://huggingface.co/docs/accelerate/basic_tutorials/troubleshooting

**Comment by IamExperimenting** [comment]
@muellerzr , I got OOM error when I used this example `https://github.com/huggingface/peft/blob/main/examples/sft/run_peft_multigpu.sh` however I reduced down to 1. I noticed when I ran the script, I noticed script printed `number of trainable parameters as 9B` for 2B gemma model.

I think we are deviating, can you please look into my script? when I fine-tune 2B gemma model with the above script in single GPU i'm getting the expected result. But when I fine-tune 2B gemma model with the above script in multi-GPU using DDP result is varying.

**Comment by IamExperimenting** [comment]
@muellerzr @SunMarc @pacman100 , can you please confirm me whether my above script is right to execute DDP in multi-gpu?

I kind a feel I'm making mistake somewhere, can you please help me?

**Comment by muellerzr** [comment]
Please try following my advice I said there, then if there are still issues please tell us. Your script looks fine. 

On your single GPU version, increase the batch size by `bs * n_gpus`. Then it will be similar. 

(And please don't ping all maintainers, when one has already responded to you)
